Index: u-boot-2024.10/bananapi_bpi-r4_emmc_env
===================================================================
--- u-boot-2024.10.orig/bananapi_bpi-r4_emmc_env
+++ u-boot-2024.10/bananapi_bpi-r4_emmc_env
@@ -2,7 +2,7 @@ ipaddr=192.168.1.1
 serverip=192.168.1.254
 loadaddr=0x50000000
 bootargs=console=ttyS0,115200n1 pci=pcie_bus_perf root=/dev/fit0 rootwait
-bootcmd=if pstore check ; then run boot_recovery ; else run boot_emmc ; fi
+bootcmd=if pstore check ; then run boot_recovery ; else run generate_hw_id; run boot_emmc ; fi
 bootconf=config-mt7988a-bananapi-bpi-r4
 bootconf_base=config-mt7988a-bananapi-bpi-r4
 bootconf_emmc=mt7988a-bananapi-bpi-r4-emmc
@@ -55,3 +55,4 @@ _init_env=setenv _init_env ; setenv _cre
 _firstboot=setenv _firstboot ; run _switch_to_menu ; run _init_env ; run boot_first
 _switch_to_menu=setenv _switch_to_menu ; setenv bootdelay 3 ; setenv bootmenu_delay 3 ; setenv bootmenu_0 $bootmenu_0d ; setenv bootmenu_0d ; run _bootmenu_update_title
 _bootmenu_update_title=setenv _bootmenu_update_title ; setenv bootmenu_title "$bootmenu_title       [33m$ver[0m"
+generate_hw_id=echo "U-Boot: Generating unique MAC..."; setexpr.l val_0 *0x11f50140; setexpr.l val_1 *0x11f50144; setexpr.l val_2 *0x11f50148; setexpr.l val_3 *0x11f5014c;setenv hw_id ${val_0}:${val_1}:${val_2}:${val_3};  echo "U-Boot: Set hw_id to ${hw_id}"; saveenv; saveenv;
Index: u-boot-2024.10/bananapi_bpi-r4_sdmmc_env
===================================================================
--- u-boot-2024.10.orig/bananapi_bpi-r4_sdmmc_env
+++ u-boot-2024.10/bananapi_bpi-r4_sdmmc_env
@@ -2,7 +2,7 @@ ipaddr=192.168.1.1
 serverip=192.168.1.254
 loadaddr=0x50000000
 bootargs=console=ttyS0,115200n1 pci=pcie_bus_perf root=/dev/fit0 rootwait
-bootcmd=if pstore check ; then run boot_recovery ; else run boot_sdmmc ; fi
+bootcmd=if pstore check ; then run boot_recovery ; else run generate_hw_id; run boot_sdmmc ; fi
 bootconf=config-mt7988a-bananapi-bpi-r4
 bootconf_sd=mt7988a-bananapi-bpi-r4-sd
 bootconf_emmc=mt7988a-bananapi-bpi-r4-emmc
@@ -64,3 +64,4 @@ _init_env=setenv _init_env ; setenv _cre
 _firstboot=setenv _firstboot ; run _switch_to_menu ; run _init_env ; run boot_first
 _switch_to_menu=setenv _switch_to_menu ; setenv bootdelay 3 ; setenv bootmenu_delay 3 ; setenv bootmenu_0 $bootmenu_0d ; setenv bootmenu_0d ; run _bootmenu_update_title
 _bootmenu_update_title=setenv _bootmenu_update_title ; setenv bootmenu_title "$bootmenu_title       [33m$ver[0m"
+generate_hw_id=echo "U-Boot: Generating unique MAC..."; setexpr.l val_0 *0x11f50140; setexpr.l val_1 *0x11f50144; setexpr.l val_2 *0x11f50148; setexpr.l val_3 *0x11f5014c;setenv hw_id ${val_0}:${val_1}:${val_2}:${val_3};  echo "U-Boot: Set hw_id to ${hw_id}"; saveenv; saveenv;
Index: u-boot-2024.10/bananapi_bpi-r4_snand_env
===================================================================
--- u-boot-2024.10.orig/bananapi_bpi-r4_snand_env
+++ u-boot-2024.10/bananapi_bpi-r4_snand_env
@@ -4,7 +4,7 @@ loadaddr=0x50000000
 bootargs=console=ttyS0,115200n1 pci=pcie_bus_perf root=/dev/fit0 rootwait ubi.block=0,fit
 bootconf=config-mt7988a-bananapi-bpi-r4
 bootconf_extra=mt7988a-bananapi-bpi-r4-emmc
-bootcmd=if pstore check ; then run boot_recovery ; else run boot_ubi ; fi
+bootcmd=if pstore check ; then run boot_recovery ; else run generate_hw_id; run boot_ubi ; fi
 bootdelay=0
 bootfile=openwrt-mediatek-filogic-bananapi_bpi-r4-initramfs-recovery.itb
 bootfile_bl2=openwrt-mediatek-filogic-bananapi_bpi-r4-snand-preloader.bin
@@ -65,3 +65,4 @@ _init_env=setenv _init_env ; run ubi_cre
 _firstboot=setenv _firstboot ; run _switch_to_menu ; run _init_env ; run boot_first
 _switch_to_menu=setenv _switch_to_menu ; setenv bootdelay 3 ; setenv bootmenu_delay 3 ; setenv bootmenu_0 $bootmenu_0d ; setenv bootmenu_0d ; run _bootmenu_update_title
 _bootmenu_update_title=setenv _bootmenu_update_title ; setenv bootmenu_title "$bootmenu_title       [33m$ver[0m"
+generate_hw_id=echo "U-Boot: Generating unique MAC..."; setexpr.l temp_val *0x11f50140; setexpr val_0 fmt %08X ${temp_val}; setexpr.l temp_val *0x11f50144; setexpr val_1 fmt %08X ${temp_val}; setexpr.l temp_val *0x11f50148; setexpr val_2 fmt %08X ${temp_val}; setexpr.l temp_val *0x11f5014c; setexpr val_3 fmt %08X ${temp_val}; setenv hw_id ${val_0}:${val_1}:${val_2}:${val_3}; echo "U-Boot: Set hw_id to ${hw_id}"; saveenv; saveenv;
