diff --git a/CMakeLists.txt b/CMakeLists.txt
index c5a08c5..e86ac3e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -139,21 +139,6 @@ function (config_without_llvm)
 endfunction ()
 
 function (config_with_llvm)
-  execute_process(COMMAND sh -c "git -C ${CMAKE_HOME_DIRECTORY}/third_party/llvm-project log -1 --format=%h"
-     RESULT_VARIABLE clang_version_status
-     OUTPUT_VARIABLE CLANG_VERSION_OUTPUT)
-
-  set(CLANG_KNOWN_GIT_COMMIT_HASH "e4f9175d2395")
-  if (clang_version_status)
-     message(WARNING "Could not get clang commit hash : Use clang git hash " ${CLANG_KNOWN_GIT_COMMIT_HASH})
-  else()
-     string(REGEX MATCH "([0-9|a-f]+)" CLANG_GIT_COMMIT_HASH ${CLANG_VERSION_OUTPUT})
-     if ("${CLANG_GIT_COMMIT_HASH}" STREQUAL ${CLANG_KNOWN_GIT_COMMIT_HASH})
-       message(STATUS "Found known git commit hash of clang (" ${CLANG_GIT_COMMIT_HASH} ") - Success")
-     else()
-       message(FATAL_ERROR "Unverified git commit hash of llvm source - ${CLANG_GIT_COMMIT_HASH}. Do not modify or update third_party/llvm-project")
-     endif()
-  endif()
 
   ### LLVM Configuration
   set (LLVM_INCLUDE_UTILS OFF)
@@ -168,8 +168,17 @@ function (config_with_llvm)
   ###
 
   find_package(absl REQUIRED)
+  set (llvm_search_paths
+    ${LLVM_PATH}
+    ${LLVM_PATH}/lib/cmake
+    ${LLVM_PATH}/lib/cmake/llvm
+    ${LLVM_PATH}/lib/cmake/clang)
+  find_package(LLVM REQUIRED CONFIG
+    PATHS ${llvm_search_paths}
+    NO_DEFAULT_PATH)
+  #set (protobuf_BUILD_TESTS OFF)
+
   add_subdirectory(third_party/glog)
-  add_subdirectory(third_party/llvm-project/llvm)
 
   add_custom_target(exclude_extlib_tests ALL 
     COMMAND rm -f ${gtest_BINARY_DIR}/CTestTestfile.cmake
@@ -176,7 +185,8 @@
     COMMAND rm -f ${absl_BINARY_DIR}/CTestTestfile.cmake)
 
   add_definitions(-DHAVE_LLVM=1)
-  include_directories(${CMAKE_HOME_DIRECTORY}
+  include_directories(${LLVM_INCLUDE_DIRS}
+    ${CMAKE_HOME_DIRECTORY}
     third_party/glog/src
     third_party/abseil
     third_party/llvm-project/llvm/include
@@ -183,6 +193,12 @@
     third_party/perf_data_converter/src/quipper
     util/regexp)
 
+  add_definitions(-DLLVM_VERSION_MAJOR=${LLVM_VERSION_MAJOR})
+  add_definitions(-DLLVM_VERSION_MINOR=${LLVM_VERSION_MINOR})
+
+  set (CMAKE_REQUIRED_INCLUDES ${LLVM_INCLUDE_DIRS})
+  set (CMAKE_REQUIRED_LIBRARIES LLVMObject)
+  set (CMAKE_REQUIRED_LIBRARIES LLVMProfileData)
   find_package(Protobuf REQUIRED)
   protobuf_generate_cpp(PERF_DATA_PROTO_CC PERF_DATA_PROTO_HDR third_party/perf_data_converter/src/quipper/perf_data.proto)
   protobuf_generate_cpp(PERF_PARSER_OPTIONS_CC PERF_PARSER_OPTIONS_HDR third_party/perf_data_converter/src/quipper/perf_parser_options.proto)
