From f662de0a64725284b4105474c73b99bb5b0b517b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Degros?= <fdegros@chromium.org>
Date: Wed, 11 Mar 2026 17:30:32 +1100
Subject: [PATCH 1/2] Accept open file descriptor as /dev/fd/nnnn

---
 libexfat/io.c | 21 +++++++++++++++++++--
 1 file changed, 19 insertions(+), 2 deletions(-)

diff --git a/libexfat/io.c b/libexfat/io.c
index bc92c7c..1ceda8a 100644
--- a/libexfat/io.c
+++ b/libexfat/io.c
@@ -61,12 +61,29 @@ static bool is_open(int fd)
 
 static int open_ro(const char* spec)
 {
-	return open(spec, O_RDONLY);
+	/* skip calling open() if fd path is provided */
+	int fd = -1;
+	int len = 0;
+	if (sscanf(spec, "/dev/fd/%u%n", &fd, &len) == 1 && len == strlen(spec)) {
+		printf("skipping open and using provided fd: %d\n", fd);
+		return fd;
+	}
+	else {
+		return open(spec, O_RDONLY);
+	}
 }
 
 static int open_rw(const char* spec)
 {
-	int fd = open(spec, O_RDWR);
+	/* skip calling open() if fd path is provided */
+	int fd = -1;
+	int len = 0;
+	if (sscanf(spec, "/dev/fd/%u%n", &fd, &len) == 1 && len == strlen(spec)) {
+		printf("skipping open and using provided fd: %d\n", fd);
+	}
+	else {
+		fd = open(spec, O_RDWR);
+	}
 #ifdef __linux__
 	int ro = 0;
 
-- 
2.53.0.851.ga537e3e6e9-goog

